### PR TITLE
Adds constants to vector

### PR DESCRIPTION
**Pull request by switzmab**
Made by Benji Switzman
- Create constant adder that will add constants to vectors
- Go through the file as well as the factory analyzer and the enum.
- Make sure implementation is fine

### CODE REVIEW

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      
+    for (int i = 1; i < data.size(); i++) {
+      try{
+        dataPoint.add(Double.toString(Double.parseDouble(data.get(i).get(v1))+a));
+        dataPoint.add(Double.toString(Double.parseDouble(data.get(i).get(v2))+b));
+        dataPoint.add(Double.toString(Double.parseDouble(data.get(i).get(v3))+c));
+        dataPoint.add(Double.toString(Double.parseDouble(data.get(i).get(v4))+d));
+        dataPoints.add(dataPoint);
+        dataPoint = new ArrayList<String>(4);
+      } catch (Exception e) {break;}
```
Comment (gr812b) [inline backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:63]: This try catch is very scary. Errors are generally thrown for a good reason, and we should handle them properly. In this case, throwing the error is probably better than ignoring it like here. @hydrowoxy is working on error handling in the backend, so it might be best to have her input on how this should be handled to make it easier for her.

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+       this.a = a;
+       this.b = b;
+       this.c = c;
+       this.d = d;
+   }
+  public void analyze() {
+
+    System.out.println("Add a constant value to a file named"+ super.inputFiles[0]+ " to make " + super.outputFiles[0]);
+    Reader r = new CSVReader(super.inputFiles[0]);
+    Writer w = new CSVWriter(super.outputFiles[0]);
```
Comment (gr812b) [inline backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:28]: I may have misled you a bit on this front, something we're looking to use going forward is to move away from using our custom CSVReader and CSVWriter and move to openCSV library's stuff. There's a whole PR open about it right now you can see here https://github.com/McMaster-Baja-Racing/Better-Data-Viewer/pull/117/files which shows a great example of changing from one to the other.

File: backend/src/main/java/com/mcmasterbaja/analyzer/AnalyzerFactory.java [hunk 3]
```diff
@@ -60,6 +60,16 @@ public static Analyzer createAnalyzer(AnalyzerParams params) {
...
 
+      case CONSTANT_ADDER:
+        if (options[3] == "" || options[2] == ""|| options[1] == "" || options[0] == "") {
+            return null;
+        }
+        double a1 = Double.parseDouble((String) options[0]);
+        double b2 = Double.parseDouble((String) options[1]);
+        double c1 = Double.parseDouble((String) options[2]);
+        double d1 = Double.parseDouble((String) options[3]);
+        return new ConstantAdderAnalyzer(inputFiles, inputColumns, outputFiles, a1, b2,c1,d1);
```
Comment (gr812b) [inline backend/src/main/java/com/mcmasterbaja/analyzer/AnalyzerFactory.java:71]: Why is this named a1, b2, c1, d1?

Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/analyzer/AnalyzerFactory.java:71]: Assuming this is because of variable names defined multiple times, this is fine I should fix this later

File: backend/src/main/java/com/mcmasterbaja/analyzer/Analyzer.java [hunk 4]
```diff
@@ -1,25 +1,26 @@
...
 import java.io.IOException;
 import java.util.List;
 
+import com.opencsv.CSVReader;
+import com.opencsv.CSVReaderBuilder;
+import com.opencsv.CSVWriter;
+import com.opencsv.CSVWriterBuilder;
+import com.opencsv.ICSVWriter;
+import com.opencsv.exceptions.CsvValidationException;
+
```
Comment (gr812b) [inline backend/src/main/java/com/mcmasterbaja/analyzer/Analyzer.java:16]: Why was this changed?

File: backend/src/main/java/com/mcmasterbaja/analyzer/Analyzer.java [hunk 5]
```diff
@@ -1,25 +1,26 @@
...
 public abstract class Analyzer {
 
   // Input and output files are arrays because some analyzers may need multiple input files
   protected String[] inputFiles;
   protected String[] inputColumns;
   protected String[] outputFiles;
-
+// Vector 4 long, add a constant to the 4 dimensions (v1+A,v2+B,v3+C,v4+D)
```
Comment (gr812b) [inline backend/src/main/java/com/mcmasterbaja/analyzer/Analyzer.java:23]: Do not include this, wrong location

File: backend/Test.csv
Comment (gr812b) [inline backend/Test.csv:1]: This should not be included

File: backend/Output.csv
Comment (gr812b) [inline backend/Output.csv:1]: This should not be included

File: README.md
Comment (gr812b) [inline README.md:1]: No changes should be made to the readme.

**Review by gr812b — CHANGES_REQUESTED** [review]
Please make sure that you have no changes on files you didn't actually work on! It makes it harder to review your code and keep track of history on files. Also please make sure not to include any test code in the final version that is not an actual test that is run during CI. 

Regarding the analyzer, I left a few comments of places to improve on. The error handling definitely needs to be looked at, but something like the reader / writer can be ignored if it's too difficult.

Finally, make sure that your PR is passing the CI! You can see it on each PR via the "Checks" tab, and it should say where it failed. In your case, the linter on the backend was not happy. You can read the `README.md` for more details on linting

File: backend/src/main/java/com/mcmasterbaja/analyzer/AnalyzerFactory.java [hunk 6]
```diff
@@ -75,6 +85,8 @@ public static Analyzer createAnalyzer(AnalyzerParams params) {
         double c = Double.parseDouble((String) options[2]);
         double d = Double.parseDouble((String) options[3]);
         return new CubicAnalyzer(inputFiles, inputColumns, outputFiles, a, b1, c, d);
+
```
Comment (t-wing11) [inline backend/src/main/java/com/mcmasterbaja/analyzer/AnalyzerFactory.java:88]: No reason to add whitespace for no reason

File: backend/src/main/java/com/mcmasterbaja/analyzer/LinearMultiplyAnalyzer.java [hunk 7]
```diff
@@ -37,7 +37,7 @@ public void analyze() {
     Reader r = new CSVReader(super.inputFiles[0]);
     Writer w = new CSVWriter(super.outputFiles[0]);
 
-    w.write(linearMultiply(r.read(), m, b));
+    w.write(linearMultiply(r.read(), m, b));
```
Comment (t-wing11) [inline backend/src/main/java/com/mcmasterbaja/analyzer/LinearMultiplyAnalyzer.java:40]: try to avoid making changes like this to random files, clutters the PR

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
+package com.mcmasterbaja.analyzer;
+
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (t-wing11) [inline backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:5]: Nitpick, don't break up the imports. Keep them as one block

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java (hunk 1 above)
Comment (t-wing11) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:63]: Seconding this

**Review by t-wing11 — COMMENTED** [review]
Code looks good, need to fix that try catch statement for sure before this gets merged.

You have lots of clutter in this PR, random files being changed, unneeded file creation. Try to avoid doing this moving forward.

Also cleanup your spacing in the code, its all over the place. Try to follow the existing code format, when adding spaces between lines, brackets, etc.

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java (hunk 2 above)
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:28]: Will need some guidance on which parts can stay vs need to be changed

Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:28]: In this [PR](https://github.com/McMaster-Baja-Racing/Better-Data-Viewer/pull/135/files#diff-bfabd33cfd2988b7b25e7a2a4b452971138a48785086fae65b0f5cec7225fd33) you can see some examples like in the image, where I swap out the CSVReader and CSVWriter (Which we define locally) for the ones that OpenCSV has. By consequence, we should also be reading files line by line, rather than loading them all at once. So instead of r.read() and then looping on it, it should loop while there exists a next line, and then performs the operations and writes them. Something like the second image here

image

image

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    return dataPoints;
+  }
+  public static void main(String[] args) {
+    /*String[] filepaths = new String[1];
+    String[] outputFiles = new String[1];
+    filepaths[0] ="C:\\Users\\benji\\OneDrive\\Desktop\\CodeShit\\BajaStuff\\Better-Data-Viewer\\backend\\Test.csv\"";
+    outputFiles[0] ="C:\\Users\\benji\\OneDrive\\Desktop\\CodeShit\\BajaStuff\\Better-Data-Viewer\\backend\\Output.csv\"";
+    ConstantAdderAnalyzer r = new ConstantAdderAnalyzer(filepaths, "4", outputFiles, 1.0, 2.0,3.0,4.0);
+    r.analyze();*/
+  }
```
[the remark this answers is not in the document]
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:76]: Should I delete the main function then?

[the remark this answers is not in the document]
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:76]: Yes, main function should be removed. Might follow up regarding writing a test for this later

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java (hunk 1 above)
Comment (hydrowoxy) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:63]: Thirding this

**Comment by JAWalmsley** [comment]
I'm a big fan of adding files individually to git with "git add [filename]" instead of just going "git add ." because it forces you to check if you actually wanted to commit the changes in that file. Saves me from accidentally pushing my autoformatter's changes or weird test files that get created. :)

File: README.md [hunk 10]
```diff
@@ -35,32 +38,32 @@ Setup complete!
...
 
 ## Linting
+
 - To run ESLint on the front end, simply download the ESLint extension (The ) in VSCode. You should then be able to run the `Auto-format file` command via the command palette (Ctrl + Shift + P).
 - To run Java Google Format, go to the repository (https://github.com/google/google-java-format) and download the latest all-deps `.jar` file. Additionally you can find it locally in the backend folder. You can then run it via the command `java -jar /path/to/google-java-format-${GJF_VERSION?}-all-deps.jar <options> [files...]`. A common option to add is`--replace` which will automatically format it.
- - To run it easier, here's a cool command (run in `backend/` directory using bash) `find ./ -name "*.java" -type f -print | xargs java -jar google-java-format-1.22.0-all-deps.jar --replace` Just make sure to check over what it changes
+- To run it easier, here's a cool command (run in `backend/` directory using bash) `find ./ -name "*.java" -type f -print | xargs java -jar google-java-format-1.22.0-all-deps.jar --replace` Just make sure to check over what it changes
 
 ## Rust Library Build/Setup (Only necessary if making library changes)
+
```
[the remark this answers is not in the document]
Comment (switzmab) [reply README.md:64]: Ok. It keeps reverting whenever I save the file

[the remark this answers is not in the document]
Comment (switzmab) [reply README.md:64]: I had formatOnSave on so that was the issue

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    
+  
+
+  //adds the variable of the line with the constant
+  private double constantAdder(double var, double constant) throws IOException {
+    return var + constant;
+
+  }
+  public static void main(String[] args) {
+
```
[the remark this answers is not in the document]
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:72]: formatOnSave is off now so should be better

File: backend/src/main/java/com/mcmasterbaja/analyzer/Analyzer.java [hunk 12]
```diff
@@ -1,25 +1,24 @@
...
+import java.io.BufferedWriter;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.List;
 
 import com.opencsv.CSVReader;
 import com.opencsv.CSVReaderBuilder;
 import com.opencsv.CSVWriter;
 import com.opencsv.CSVWriterBuilder;
```
[the remark this answers is not in the document]
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/Analyzer.java:12]: When I try to remove those it causes an error

File: README.md
Comment (switzmab) [reply README.md:1]: Should be good now

File: backend/Output.csv
Comment (switzmab) [reply backend/Output.csv:1]: Removed

File: backend/Test.csv
Comment (switzmab) [reply backend/Test.csv:1]: Removed

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java (hunk 1 above)
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java]: try catch removed

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java (hunk 8 above)
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java]: fixed

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java (hunk 2 above)
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java]: good now

File: backend/src/main/java/com/mcmasterbaja/model/AnalyzerType.java [hunk 13]
```diff
@@ -10,7 +10,7 @@ public enum AnalyzerType {
   RDP_COMPRESSION,
   ROLL_AVG,
   SGOLAY,
-  SPLIT;
+  SPLIT, CONSTANT_ADDER;
```
[the remark this answers is not in the document]
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/model/AnalyzerType.java]: done

File: backend/src/main/java/com/mcmasterbaja/analyzer/Analyzer.java (hunk 5 above)
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/Analyzer.java]: Deleted it

File: backend/src/main/java/com/mcmasterbaja/analyzer/LinearMultiplyAnalyzer.java (hunk 7 above)
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/analyzer/LinearMultiplyAnalyzer.java:40]: This convo has not been resolved

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (gr812b) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java]: This was not resolved

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+        String newD = Double.toString(constantAdder(oldD, d));
+        w.writeNext(new String[] {newA, newB, newC, newD});
+      }
+      r.close();
+      w.close();
+    }
+  }
+
+  // adds the variable of the line with the constant
+  private double constantAdder(double var, double constant) throws IOException {
```
Comment (gr812b) [inline backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:64]: Why does this throw an IOException?

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+    return var + constant;
```
Comment (gr812b) [inline backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:65]: A function simply to add a value is a bit redundant

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+  public void analyze() throws IOException, CsvValidationException {
+
+    System.out.println(
+        "Add a constant value to a file named"
+            + super.inputFiles[0]
+            + " to make "
+            + super.outputFiles[0]);
+    CSVReader r = getReader(super.inputFiles[0]);
+    ICSVWriter w = getWriter(super.outputFiles[0]);
```
Comment (gr812b) [inline backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:37]: Nitpicky but every other analyzer calls these `reader` and `writer` respectively, may be nice to stay consistent

**Review by gr812b — CHANGES_REQUESTED** [review]
Almost there! Just a few things to clean up

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java (hunk 16 above)
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:37]: sure will change

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java (hunk 14 above)
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:64]: Forgot to remove... when I first edited I had some read and write in there

File: backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java (hunk 15 above)
Comment (switzmab) [reply backend/src/main/java/com/mcmasterbaja/analyzer/ConstantAdderAnalyzer.java:65]: Function no more!

**Review by gr812b — APPROVED** [review]
